text/0009-BundleTracker-integration.md

```diff
@@ -0,0 +1,588 @@
+
+**Responsibilities:**
+
+1. Define the inherited customizer method wrappers as calling the user `BundleTrackerCustomizer` methods.
+2. Define the top-level `BundleChanged` function to be called by the `BundleListener`. This method contains the business logic of deciding which customizer to call based on the current state of the tracking map, the state mask, and the intercepted `BundleEvent`.
+3. Keep a tracking map (implementation handled by parent class `BundleAbstractTracked`).
+4. Provide a wait function so the `BundleTracker` can delay closing until the customizer methods return.
+
+**Concurrency Note:**
+
+This object is stored as an atomic shared pointer in `BundleTrackerPrivate` to provide atomic access to `TrackedBundle` data. It is also guaranteed that the customizer wrappers are called while not holding any locks.
+
+### Communication Diagrams
+
+#### Opening BundleTracker
+![Open Diagram](0009-BundleTracker-integration/open_diagram.svg)
+
+#### Untracked Bundle Enters Tracked State
+![Adding](0009-BundleTracker-integration/addbundle_diagram.svg)
+
+#### Tracked Bundle Enters Tracked State
+![Modified](0009-BundleTracker-integration/modbundle_diagram.svg)
+
+#### Tracked Bundle Enters Untracked State
+![Removed](0009-BundleTracker-integration/rembundle_diagram.svg)
+
+#### Bundle Removed by User
+![Manually Removed](0009-BundleTracker-integration/manrembundle_diagram.svg)
+
+#### Closing BundleTracker
+![Closed](0009-BundleTracker-integration/close_diagram.svg)
+
+## How we teach this
+****
+The `BundleTracker` concept is best presented as a parallel to the `ServiceTracker`, which already exists and has the same customizer and extender patterns. A relevant difference between the constructs besides what is being tracked, is the method of filtering. Services offer class names and filters to target certain services. On the contrary, the `BundleTracker` uses a bit mask for `Bundle` states. This difference should be explained in comparing the `BundleTracker` to the `ServiceTracker`.
+
+Given the acceptance of this RFC, it would be worthwhile to include an example of the `BundleTracker` being used in a CppMicroServices example. It could also be included in a "Best Practices" document to urge existing and new users to use the `BundleTracker` implementation instead of creating their own solutions to the tracking problem.
+
+Since the `BundleTracker` is not a feature exposed in a beginner's workflow, it is not urgent to include it in a "Getting Started" guide.
+
+## Drawbacks
+****
+
+There will be inconsistency in API and implementation details between the `BundleTracker` and the `ServiceTracker`, which could create some confusion for users. Additionally, since the `BundleTracker` is built upon a *synchronous* event listener, it allows the user to slow the framework by doing non-trivial work in callbacks.
```

Comment:
The servicetracker is also built upon a synchronous listener, so at least bundletracker is at parity with servicetracker in this regard.

Reply:
This parallel is made explicit in the RFC